src/dbnode/storage/series/buffer.go

```diff
@@ -268,65 +385,106 @@ func bucketTick(now time.Time, b *dbBuffer, idx int, start time.Time) int {
 
 func (b *dbBuffer) DrainAndReset() drainAndResetResult {
 	// Avoid capturing any variables with callback
-	mergedOutOfOrder := b.computedForEachBucketAsc(computeAndResetBucketIdx, bucketDrainAndReset)
+	mergedOutOfOrder := b.computedForEachBucketAsc(computeAndResetBucketIdx, bucketTypeRealTime, bucketDrainAndReset)
 	return drainAndResetResult{
 		mergedOutOfOrderBlocks: mergedOutOfOrder,
 	}
 }
 
-func bucketDrainAndReset(now time.Time, b *dbBuffer, idx int, start time.Time) int {
+func bucketDrainAndReset(now time.Time, b *dbBuffer, id bucketID, start time.Time) int {
 	mergedOutOfOrderBlocks := 0
+	bucket := b.bucketAtIdx(id)
 
-	if b.buckets[idx].needsDrain(now, start) {
-		// Rotate the buffer to a block, merging if required
-		result, err := b.buckets[idx].discardMerged()
-		if err != nil {
+	if bucket.needsDrain(now, start) {
+		mergedOutOfOrderBlocks += b.bucketDrain(now, id, start)
+	}
+
+	if bucket.needsReset(now, start) {
+		bucket.resetTo(start, true)
+	}
+
+	if bucket.isStale(now) {
+		b.removeBucket(id.key)
```

Comment:
Is it worth adding a sanity check to remove bucket to emit an error log if its called on a bucket with more than 1 write or one that is not drained? Kind of like an invariant violated assert?